### PR TITLE
Correct WSHP constant fan operating mode

### LINKED ISSUE #10486: ZoneHVAC:WaterToAirHeatPump air flow control issue
Issue overview
--------------
The attached [example file](https://github.com/TiejunWu/EnergyPlus-IDF-Files/blob/master/ZoneVSWSHP_wDOAS1.idf) is based on the EnergyPlus installed example file "ZoneVSWSHP_wDOAS.idf". The only change I made was to change the zone heat pump fan operating mode control from cycling to constant. 

Based on the way the zone water to air heat pump is specified, my understanding is that it should run with the sized air flow all the time.  From sizing report, I see that the zone terminal fan is sized at 0.939 m3/s.  So when the fan operating model is constant, I would expect the flow rate to be a straight line at 0.939 m3/s.

image

However, when I look at the designday simulation results. The fan air flow seems to be coordinated with the coil in some way.  When there is no cooling or heating, the air flow is at 0.7 m3/s. But when the cooling coil is running, the fan air flow is at 0.338 m3/s. I am not sure how the program decided to run the fan at those flow rates. In the water to air heat pump input, there is no "UnitarySystemPerformance:Multispeed" object referenced. 

image

image

image

### Details
Some additional details for this issue (if relevant):
 - Platform (Windows 10 64 Bit)
 - Version of EnergyPlus (EnergyPlus 24.1)
 - Unmethours link or helpdesk ticket number: N/A

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Defect file added ([ZoneVSWSHP_wDOAS1.idf](https://github.com/TiejunWu/EnergyPlus-IDF-Files/blob/master/ZoneVSWSHP_wDOAS1.idf))
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes #10486
 - ZoneHVAC:WaterToAirHeatPump heating mode does not honor a constant fan operating mode schedule when a variable or multi-speed heating coil is used. When operating in cooling mode the constant fan operating mode works correctly.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/UnitarySystem.cc [hunk 1]
```diff
@@ -10565,12 +10558,9 @@ namespace UnitarySystems {
                         if (HeatSpeedNum == 0) {
                             state.dataUnitarySystems->CompOffMassFlow = this->MaxNoCoolHeatAirMassFlow;
                             state.dataUnitarySystems->CompOffFlowRatio = this->m_NoLoadAirFlowRateRatio;
-                        } else if (HeatSpeedNum == 1) {
+                        } else {
                             state.dataUnitarySystems->CompOffMassFlow = this->m_HeatMassFlowRate[HeatSpeedNum];
                             state.dataUnitarySystems->CompOffFlowRatio = this->m_HeatMassFlowRate[HeatSpeedNum];
-                        } else {
-                            state.dataUnitarySystems->CompOffMassFlow = this->m_HeatMassFlowRate[HeatSpeedNum - 1];
-                            state.dataUnitarySystems->CompOffFlowRatio = this->m_MSHeatingSpeedRatio[HeatSpeedNum - 1];
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:10573]: This was the problem in the defect file. In heating mode, when constant fan is used, the OFF mass flow rate should be the same as the ON mass flow rate.

File: tst/EnergyPlus/unit/UnitarySystem.unit.cc [hunk 2]
```diff
@@ -5085,16 +5085,16 @@ Schedule:Compact,
     state->dataUnitarySystems->HeatingLoad = true;
     state->dataUnitarySystems->CoolingLoad = false;
     thisSys.setOnOffMassFlowRate(*state, OnOffAirFlowRatio, PartLoadRatio);
-    EXPECT_EQ(0.6, state->dataUnitarySystems->CompOffMassFlow);
+    EXPECT_EQ(1.2, state->dataUnitarySystems->CompOffMassFlow);
     EXPECT_EQ(1.2, state->dataUnitarySystems->CompOnMassFlow);
-    EXPECT_EQ(0.6, state->dataHVACGlobal->MSHPMassFlowRateLow);
+    EXPECT_EQ(1.2, state->dataHVACGlobal->MSHPMassFlowRateLow);
     EXPECT_EQ(1.2, state->dataHVACGlobal->MSHPMassFlowRateHigh);
```
Comment (rraustad) [inline tst/EnergyPlus/unit/UnitarySystem.unit.cc:5091]: This unit test should have noticed the issue with on and off air flow. This is correct now.

**Comment by rraustad** [comment]
Top row is from PackagedTerminalAirConditionerVSAS and bottom row is PackagedTerminalHeatPumpVSAS example files that show big diffs. The fan/airflow is now constant as expected. The VS coils model constant air flow at speed which may not be the best choice for constant fan operating mode but that's what these coils do now.

image

File: src/EnergyPlus/UnitarySystem.cc [hunk 3]
```diff
@@ -10565,12 +10558,9 @@ namespace UnitarySystems {
                         if (HeatSpeedNum == 0) {
                             state.dataUnitarySystems->CompOffMassFlow = this->MaxNoCoolHeatAirMassFlow;
                             state.dataUnitarySystems->CompOffFlowRatio = this->m_NoLoadAirFlowRateRatio;
-                        } else if (HeatSpeedNum == 1) {
+                        } else {
                             state.dataUnitarySystems->CompOffMassFlow = this->m_HeatMassFlowRate[HeatSpeedNum];
```
Comment (mjwitte) [inline src/EnergyPlus/UnitarySystem.cc]: Why doesn't this block set `CompOnMassFlow`? Throughout this function some block set both `CompOnMassFlow` and `CompOffMassFlow` but some just set the off values.

Comment (rraustad) [reply src/EnergyPlus/UnitarySystem.cc]: Sorry, I did see this but got sidetracked and forgot to answer. Look at line 10536 above. This block of code is only setting the CompOffMassFlow based on whether or not the fan is cycling. The CompOnMassFlow gets set regardless of fan operation elsewhere (hopefully!).

Comment (mjwitte) [reply src/EnergyPlus/UnitarySystem.cc]: Ok, I see. There's a "default" `CompOnMassFlow` at the top of this section, and then some cases (maybe just 1) need to override that, but not all. I'll update this branch with develop and do a final review pass.

**Comment by mjwitte** [comment]
So, there are quite a few diffs. And some of the error diffs don't look good. Here's a couple of the error diff files (from a local regression run).
[UnitarySystem_MultiSpeedDX-eplusout.err.diff.txt](https://github.com/user-attachments/files/15830038/UnitarySystem_MultiSpeedDX-eplusout.err.diff.txt)
[ASHRAE901_RestaurantFastFood_STD2019_Denver-eplusout.err.diff.txt](https://github.com/user-attachments/files/15830046/ASHRAE901_RestaurantFastFood_STD2019_Denver-eplusout.err.diff.txt)

**Comment by rraustad** [comment]
@mjwitte I almost agreed with you. The diffs in UnitarySystem_MultiSpeedDX do look disturbing with coil control failed that were not there before. The diffs in ASHRAE901_RestaurantFastFood_STD2019_Denver are of the same type but with a change in number of warnings. Then I looked at the diffs on this branch. What is this? Are we looking at the right diffs?

image

**Comment by rraustad** [comment]
OK, regardless of the CI diffs I'll try to run these 2 locally to see if something stands out.

**Comment by rraustad** [comment]
Something does look different. These figures show that the WSHP air flow is higher now during the cooling period yet the zone temperature does not meet set point as it does in develop. Not sure what is causing this.

image

**Comment by rraustad** [comment]
Here's what I see in UnitarySystem_MultiSpeedDX. Where UnitarySystem determines what speed to operate at to meet the load, the 1st stage max capacity is 2054 W. The cooling load is slightly larger (2120 W) so it chooses 2nd stage. Then when the program tries to find a speed ratio at stage 2, it shows the capacity at speed = 2 with speed ratio = 0 is 2576 W. Of course SolveRoot fails with -2. So why is the top end of speed = 1 different from the bottom end of speed = 2? Perf curves?

image

**Comment by rraustad** [comment]
When the Coil:Cooling:DX coil has NumSpeed>1 the UnitarySystem calculates the speed = 1 result at the speed = 1 air flow. Then it calculates the speed = 2 result at the speed = 2 air flow. When this coil operates at speed = 2 it calculates it's output at speed 2 and then evaluates the output at speed = 1, however, it uses MSHPMassFlowRateLow as the air flow at speed = 1, and in the case of constant fan MSHPMassFlowRateLow = MSHPMassFlowRateHigh = air flow at speed = 2. When the coil evaluates the result at speed = 1 at this higher air flow the coil "rated" capacity is the same, the modifiers are roughly the same, but the SHR changes from 0.72 at the lower air flow to SHR =1 at this higher air flow so the outlet temperature is no longer proportional to air flow and hence the "sensible" output at speed = 1 and CyclingRatio = 1 is different than output at speed = 2 and SpeedRatio = 0. I'm not quite sure what to do here other than check for this capacity while choosing speed num and exiting out if this scenario exists (lower speed at PLR = 1 is lower than load and higher speed at PLR = 0 is greater than load).

    if ((speedNum > 1) && (speedRatio < 1.0) && !singleMode) {

        // If multispeed, evaluate next lower speed using PLR, then combine with high speed for final outlet conditions
        auto &lowerspeed(this->speeds[max(speedNum - 2, 0)]);
        lowerspeed.AirMassFlow = state.dataHVACGlobal->MSHPMassFlowRateLow * lowerspeed.active_fraction_of_face_coil_area;

**Comment by rraustad** [comment]
I guess the other thing I could do is set MSHPMassFlowRateLow to the lower speed air flow rate but when using constant fan mode does this make sense? Doing this should make the capacity at speed = 1 and CyclingRatio = 1 nearly the same as the capacity at speed = 2 and SpeedRatio = 0. However, I don't think this makes sense since if you raise the air flow going to a higher speed then the performance of the coil at the lower speed will change just as I see with the change in coil SHR. I'm stuck with these 2 results and having to make a decision on what to do in the parent.

**Comment by rraustad** [comment]
This last commit does not change the answer this branch gives but does eliminate the SolveRoot error that showed up because when SolveRoot failes with -2 it returns the lower bound which is PLR=0.

I also did not understand why the zone temp was different now, it's because higher speed air flow also gives more sensible capacity from the lower speed at higher air flow so the load is exceeded.

image

**Comment by rraustad** [comment]
I am wondering if I made a bad assumption with VS coils. If the fan operating mode is constant does that mean that the fan cannot change speeds, or just that the fan cannot turn off? The original issue suggests that with constant fan the fan should operate at the max air flow rate. This certainly was not happening before and is not happening now. I think a Plan B is warranted.

**Comment by Myoldmopar** [comment]
> I think a Plan B is warranted.

@rraustad do you intend on still using this PR then?  I'm just going through the list this morning and doing spot checks.

**Comment by rraustad** [comment]
@Myoldmopar I am leaning towards closing this branch.

**Comment by rraustad** [comment]
Closing as not a good fix for this issue

**Pull request closed without merging**